01-intro.Rmd

```diff
@@ -8,3 +8,284 @@ ottrpal::set_knitr_image_path()
 Welcome!
 
 - Review of basic WDL syntax
+- How to use input JSONs
+- (optional) Installing Docker and miniwdl
+- How to run simple workflows locally
+
+
+## Review of basic WDL syntax
+A WDL workflow consists of at least one task.
+
+<!-- resources/basic_01.wdl -->
+```
+version 1.0
+
+task do_something {
+    command <<<
+        exit 0
+    >>>
+}
+
+workflow my_workflow {
+    call do_something
+}
+```
+
+A workflow, and the tasks it calls, generally has inputs.
+
+<!-- resources/basic_02.wdl -->
+```
+version 1.0
+
+task do_something {
+    input {
+        File fastq
+    }
+    command <<<
+        exit 0
+    >>>
+}
+
+workflow my_workflow {
+    input {
+        File fq
+    }
+    call do_something {
+        input:
+            fastq = fq
+    }
+}
+```
+
+To access a task-level input variable in a task's command section, it is usually referenced using ~{this} notation. To access a workflow-level variable in a workflow, it is referenced just by its name without any special notation. To access a workflow-level variable in a task, it must be passed into the task as an input.  
+
+<!-- resources/basic_03.wdl -->
+```
+version 1.0
+
+task do_something {
+    input {
+        File fastq
+        String basename_of_fq
+    }
+    command <<<
+        echo "First ten lines of ~{basename_of_fq}: "
+        head ~{fastq}
+    >>>
+}
+
+workflow my_workflow {
+    input {
+        File fq
+    }
+    
+    String basename_of_fq = basename(fq)
+    
+    call do_something {
+        input:
+            fastq = fq,
+            basename_of_fq = basename_of_fq
+    }
+}
+```
+
+Tasks and workflows also typically have outputs. The task-level outputs can be accessed by the workflow or any subsequent tasks. The workflow-level outputs represent the final output of the overall workflow.
+
+<!-- resources/basic_04.wdl -->
+```
+version 1.0
+
+task do_something {
+    input {
+        File fastq
+        String basename_of_fq
+    }
+    command <<<
+        echo "First ten lines of ~{basename_of_fq}: " >> output.txt
+        head ~{fastq} >> output.txt
+    >>>
+    output {
+        File first_ten_lines = "output.txt"
```

Comment:
Hardcoding output file names is sometimes a good idea, and this is a little bit easier as a first example. We should definitely bring this up later though.

For example, a workflow is trying to create 100 VCFs from 100 fastqs should indeed have those VCFs end up with filenames that are based on the fastqs -- otherwise you end up with 100 VCFs with the same filename. That being said, a workflow that takes in 100 VCFs and generates some sort of single-file report on their contents should probably have a hardcoded name for that report (or a name that the user specifies, although that can create issues if the user picks characters filesystems tend not to like in filenames like `/`).

Reply:
Ok that's fair! Let's bring it up later then.